[Docs]: Add advanced_guides/engine.md (#454)
open-mmlab/mmselfsup: docs/en/advanced_guides/engine.md

|                [EMAHook](https://github.com/open-mmlab/mmengine/blob/main/mmengine/hooks/ema_hook.py)                 |              apply Exponential Moving Average (EMA) on the model during training.              | NORMAL (50)  |
|         [EmptyCacheHook](https://github.com/open-mmlab/mmengine/blob/main/mmengine/hooks/empty_cache_hook.py)         |            release all unoccupied cached GPU memory during the process of training.            | NORMAL (50)  |
|        [SyncBuffersHook](https://github.com/open-mmlab/mmengine/blob/main/mmengine/hooks/sync_buffer_hook.py)         | synchronize model buffers such as running_mean and running_var in BN at the end of each epoch. | NORMAL (50)  |
| [NaiveVisualizationHook](https://github.com/open-mmlab/mmengine/blob/main/mmengine/hooks/naive_visualization_hook.py) |               Show or Write the predicted results during the process of testing.               | LOWEST (100) |

### Hooks implemented in MMSelfsup

Some hooks have been already implemented in MMSelfsup, they are:

> - [DeepClusterHook](https://github.com/open-mmlab/mmselfsup/blob/dev-1.x/mmselfsup/engine/hooks/deepcluster_hook.py)

Reviewer: Better to use API path instead of github link, the same comments about below

`loss_lambda` is loss weight for the single and dense contrastive loss. Defaults to 0.5.

```python
losses = dict()
losses['loss_single'] = loss_single * (1 - self.loss_lambda)
losses['loss_dense'] = loss_dense * self.loss_lambda
```

`DenseCLHook` is implemented as follows:

```python
...
@HOOKS.register_module()
class DenseCLHook(Hook):
...
    def before_train_iter(self,
                          runner,
                          batch_idx: int,
                          data_batch: Optional[Sequence[dict]] = None) -> None:
...
        cur_iter = runner.iter
        if cur_iter >= self.start_iters:
            get_model(runner.model).loss_lambda = self.loss_lambda
        else:
            get_model(runner.model).loss_lambda = 0.

```

If the hook is already implemented in MMEngine or MMSelfsup, you can directly modify the config to use the hook as below

```python
custom_hooks = [
    dict(type='MMEngineHook', a=a_value, b=b_value, priority='NORMAL')
]
```

such as using `EMAHook`, start_iters is 500:

```python
custom_hooks = [
    dict(type='DenseCLHook', start_iters=500)
]
```

## Optimizer

> 

Reviewer: Add one description here to we introduce `Optimizer` section through 3 different parts

### Optimizer

> #### Customize optimizer supported by Pytorch

Reviewer 2: ```suggestion
#### Customize optimizer supported by PyTorch
```

> We already support to use all the optimizers implemented by PyTorch, see `mmengine/optim/optimizer/builder.py`. To use and modify them, please change the `optimizer` field of config files.

Reviewer: ```suggestion
We have already supported all the optimizers implemented by PyTorch, see `mmengine/optim/optimizer/builder.py`. To use and modify them, please change the `optimizer` field of config files.
```

For example, if you want to use SGD, the modification could be as the following.

```python
optimizer = dict(type='SGD', lr=0.0003, weight_decay=0.0001)
```

To modify the learning rate of the model, just modify the `lr` in the config of optimizer. You can also directly set other arguments according to the [API doc](https://pytorch.org/docs/stable/optim.html?highlight=optim#module-torch.optim) of PyTorch.

For example, if you want to use `Adam` with the setting like `torch.optim.Adam(params, lr=0.001, betas=(0.9, 0.999), eps=1e-08, weight_decay=0, amsgrad=False)` in PyTorch, the config should looks like:

```python
optimizer = dict(type='Adam', lr=0.001, betas=(0.9, 0.999), eps=1e-08, weight_decay=0, amsgrad=False)
```

#### Parameter-wise configuration

> Some models may have some parameter-specific settings for optimization, for example, no weight decay to the BatchNorm layer and the bias in each layer. To finely configure them, we can use the `paramwise_options` in optimizer.

Reviewer: ```suggestion
Some models may have some parameter-specific settings for optimization, for example, no weight decay to the BatchNorm layer and the bias in each layer. To finely configure them, we can use the `paramwise_cfg` in optimizer.
```

> For example, if we do not want to apply weight decay to the parameters of `BatchNorm` or `GroupNorm`, and the bias in each layer, we can use following config file:

Reviewer: Take MAE config as an example and correct the code snippet

```python
optimizer = dict(
    type=...,
    lr=...,
    paramwise_options={
        '(bn|gn)(\\d+)?.(weight|bias)':
        dict(weight_decay=0.),
        'bias': dict(weight_decay=0.)
    })
```

#### Implemented optimizers in MMSelfsup

> - [LARS](https://github.com/open-mmlab/mmselfsup/blob/dev-1.x/mmselfsup/engine/optimizers/lars.py)

Reviewer: ```suggestion
- [LARS](mmselfsup.engine.optimizers.LARS)
```

In addition to optimizers implemented by PyTorch, we also implement a customized [LARS](https://arxiv.org/abs/1708.03888) in `mmselfsup/engine/optimizers/lars.py`. It implements layer-wise adaptive rate scaling for SGD.

```python
optimizer = dict(type='LARS', lr=4.8, momentum=0.9, weight_decay=1e-6)
```

> ### Optimizer Wrapper

Reviewer 2: ```suggestion
### Optimizer wrapper
```

```python
optimizer = dict(type='SGD', lr=0.01, momentum=0.9, weight_decay=0.0001)
optim_wrapper = dict(
    type='OptimWrapper',
	optimizer=optimizer,
    clip_grad=dict(
        max_norm=0.2,
        norm_type=2))
# norm_type: type of the used p-norm, here norm_type is 2.
```

If `clip_grad` is not None, it will be the arguments of `torch.nn.utils.clip_grad.clip_grad_norm_()`.

#### Gradient accumulation

When there is not enough computation resource, the batch size can only be set to a small value, which may degrade the performance of model. Gradient accumulation can be used to solve this problem.

Here is an example:

```python
train_dataloader = dict(batch_size=64)
optim_wrapper = dict(
    type='OptimWrapper',
    optimizer=optimizer,
    accumulative_counts=4)
```

Indicates that during training, back-propagation is performed every 4 iters. And the above is equivalent to:

```python
train_dataloader = dict(batch_size=256)
optim_wrapper = dict(
    type='OptimWrapper',
    optimizer=optimizer,
    accumulative_counts=1)
```

> #### automatic mixed precision training

Reviewer 2: ```suggestion
#### Automatic mixed precision(AMP) training
```

```python
optimizer = dict(type='SGD', lr=0.01, momentum=0.9, weight_decay=0.0001)
optim_wrapper = dict(type='AmpOptimWrapper', optimizer=optimizer)
```

The default setting of `loss_scale` of `AmpOptimWrapper` is `dynamic`.

### Constructor

The constructor aims to build optimizer, optimizer wrapper and customize hyper-parameters of different layers. The key `paramwise_cfg` of `optim_wrapper` in configs controls this customization.

> #### Implemented constructors in MMSelfsup

Reviewer 2: ```suggestion
#### Constructors implemented in MMSelfsup
```

> - [LearningRateDecayOptimWrapperConstructor](https://github.com/open-mmlab/mmselfsup/blob/dev-1.x/mmselfsup/engine/optimizers/layer_decay_optim_wrapper_constructor.py#L64)

Reviewer: ```suggestion
- [LearningRateDecayOptimWrapperConstructor](mmselfsup.engine.optimizers.LearningRateDecayOptimWrapperConstructor)
```

Hooks only work after being registered into the runner. At present, hooks are mainly divided into two categories:

- default hooks

Those hooks are registered by the runner by default. Generally, they fulfill some basic functions, and have default priority, you don't need to modify the priority.

- custom hooks

The custom hooks are registered through custom_hooks. Generally, they are hooks with enhanced functions. The priority needs to be specified in the configuration file. If you do not specify the priority of the hook, it will be set to 'NORMAL' by default.

**Priority list**:

|      Level      | Value |
| :-------------: | :---: |
|     HIGHEST     |   0   |
|    VERY_HIGH    |  10   |
|      HIGH       |  30   |
|  ABOVE_NORMAL   |  40   |
| NORMAL(default) |  50   |
|  BELOW_NORMAL   |  60   |
|       LOW       |  70   |
|    VERY_LOW     |  90   |
|     LOWEST      |  100  |

The priority determines the execution order of the hooks. Before training, the log will print out the execution order of the hooks at each stage to facilitate debugging.

> ### default hooks

Reviewer 2: ```suggestion
### Default hooks
```

> Some common hooks are not registered through `custom_hooks`, they are

Reviewer 2: ```suggestion
The following common hooks are already reigistered by [default][(https://github.com/open-mmlab/mmselfsup/blob/dev-1.x/configs/selfsup/_base_/default_runtime.py#L3), which is implemented through [`register_default_hooks`](https://github.com/open-mmlab/mmengine/blob/main/mmengine/runner/runner.py#L1750) in MMEngine
```

|       [IterTimerHook](https://github.com/open-mmlab/mmengine/blob/main/mmengine/hooks/iter_timer_hook.py)       |             log the time spent during iteration.             |    NORMAL (50)    |
| [DistSamplerSeedHook](https://github.com/open-mmlab/mmengine/blob/main/mmengine/hooks/sampler_seed_hook.py) |         ensure distributed Sampler shuffle is active         |    NORMAL (50)    |
|          [LoggerHook](https://github.com/open-mmlab/mmengine/blob/main/mmengine/hooks/logger_hook.py)          | collect logs from different components of ``Runner`` and write them to terminal, JSON file, tensorboard and wandb .etc. | BELOW_NORMAL (60) |
|  [ParamSchedulerHook](https://github.com/open-mmlab/mmengine/blob/main/mmengine/hooks/param_scheduler_hook.py)  | update some hyper-parameters in optimizer, e.g., learning rate and momentum. |     LOW (70)      |
|      [CheckpointHook](https://github.com/open-mmlab/mmengine/blob/main/mmengine/hooks/checkpoint_hook.py)      |                save checkpoints periodically.                |   VERY_LOW (90)   |

> ### Use other implemented hooks

Reviewer 2: ```suggestion
### Common Hooks implemented in MMEngine
```

Some hooks have been already implemented in MMEngine, they are:

|                            Hooks                             |                            Usage                             |   Priority   |
| :----------------------------------------------------------: | :----------------------------------------------------------: | :----------: |
| [EMAHook](https://github.com/open-mmlab/mmengine/blob/main/mmengine/hooks/ema_hook.py) | apply Exponential Moving Average (EMA) on the model during training. | NORMAL (50)  |
| [EmptyCacheHook](https://github.com/open-mmlab/mmengine/blob/main/mmengine/hooks/empty_cache_hook.py) | release all unoccupied cached GPU memory during the process of training. | NORMAL (50)  |
| [SyncBuffersHook](https://github.com/open-mmlab/mmengine/blob/main/mmengine/hooks/sync_buffer_hook.py) | synchronize model buffers such as running_mean and running_var in BN at the end of each epoch. | NORMAL (50)  |
| [NaiveVisualizationHook](https://github.com/open-mmlab/mmengine/blob/main/mmengine/hooks/naive_visualization_hook.py) | Show or Write the predicted results during the process of testing. | LOWEST (100) |

> ### Implemented hooks in MMSelfsup

Reviewer 2: ```suggestion
### Hooks implemented in MMSelfsup
```
